# **REDES NEURAIS ARTIFICIAIS: CLASSIFICAÇÃO**

Este projeto tem por objetivo desenvolver um algoritmo de Machine Learning para prever se a pessoa possui um cancer benigno ou maligno

Os dados foram extraídos do site do Kaggle:

https://www.kaggle.com/uciml/breast-cancer-wisconsin-data

In [1]:
import numpy as np
import pandas as pd

In [2]:
df = pd.read_csv('cancer.csv',
                    sep=',', encoding='iso-8859-1')

In [3]:
df.head()

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0         M        17.99         10.38          122.80     1001.0   
1         M        20.57         17.77          132.90     1326.0   
2         M        19.69         21.25          130.00     1203.0   
3         M        11.42         20.38           77.58      386.1   
4         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280        198.0000              0.10430   

   symmetry_mean  ...  radius_worst  texture_worst  perimeter_worst  \
0         0.2419  ...         25.38          17.33           184.60   
1         0.1812  ...         24.99          23.41           158.80   
2         0.2069  ...         23.57          25.53           152.50   
3         0.2597  ...         14.91          26.50            98.87   
4         0.1809  ...         22.54          16.67           152.20   

   area_worst  smoothness_worst  compactness_worst  concavity_worst  \
0      2019.0            0.1622             0.6656           0.7119   
1      1956.0            0.1238             0.1866           0.2416   
2      1709.0            0.1444             0.4245           0.4504   
3       567.7            0.2098             0.8663           0.6869   
4      1575.0            0.1374           205.0000           0.4000   

   concave points_worst  symmetry_worst  fractal_dimension_worst  
0                0.2654          0.4601                  0.11890  
1              186.0000        275.0000                  0.08902  
2              243.0000          0.3613                  0.08758  
3                0.2575          0.6638                173.00000  
4                0.1625          0.2364                  0.07678  

[5 rows x 31 columns]

## **LEGENDA**

id - Número de ID
diagnosis (M = malignant, B = benign) - Diagnóstico (M = maligno, B = benigno)

a) radius (mean of distances from center to points on the perimeter) - raio (média das distâncias do centro aos pontos no perímetro)

b) texture (standard deviation of gray-scale values) - textura (desvio padrão dos valores de escala de cinza)

c) perimeter - perímetro

d) area - área

e) smoothness (local variation in radius lengths) - suavidade (variação local nos comprimentos do raio)

f) compactness (perimeter^2 / area - 1.0) - compacidade (perímetro² / área - 1.0)

g) concavity (severity of concave portions of the contour) - concavidade (gravidade das porções côncavas do contorno)

h) concave points (number of concave portions of the contour) - pontos côncavos (número de porções côncavas do contorno)

i) symmetry - simetria

j) fractal dimension ("coastline approximation" - 1) - dimensão fractal ("aproximação da linha costeira" - 1)


A média, o erro padrão e o "pior" ou maior (média dos três maiores valores) dessas características foram calculados para cada imagem, resultando em 30 características. Por exemplo, o campo 3 é o Mean Radius - Raio Médio, o campo 13 é o Radius SE - Erro Padrão do Raio, e o campo 23 é o Worst Radius - Pior Raio.


Todos os valores das características são registrados com quatro dígitos significativos.

Valores ausentes de atributos: nenhum

Distribuição das classes: 357 benignos, 212 malignos


In [4]:
df.shape

(569, 31)

**Resultados**

Naive Bayes = 94,22% (treino e teste) - 375 acertos e 93.47% (validação cruzada) - previsores_esc

SVM = 90,95% - 362 acertos e 97,88% (validação cruzada) : previsores-esc - SVC(kernel='rbf', random_state=1, C = 2)

Regressão logística = 96,23% - 383 acertos e 98,06% (validação cruzada) : previsores_esc - LogisticRegression(random_state=1, max_iter=600, penalty="l2", tol=0.0001, C=1,solver="lbfgs")

KNN = 94,97% - 378 acertos e 96,65% (validação cruzada) : previsores_esc ou previsores_esc - KNeighborsClassifier(n_neighbors=7, metric='minkowski', p = 1)

Árvore de decisão = 93,71% (treino e teste) - 373 acertos e 90.34% (validação cruzada): previsores_esc - DecisionTreeClassifier(criterion='entropy', random_state = 0, max_depth=3)

**Random Forest = 98,99% - 394 acertos e 95,76% (validação cruzada) : previsores - RandomForestClassifier(n_estimators=150, criterion='entropy', random_state = 0, max_depth=4)**

XGboost = 100% - 398 acertos e 96,65% (validação cruzada) : previsores - XGBClassifier(max_depth=2, learning_rate=0.05, n_estimators=250, objective='binary:logistic', random_state=3)

LightGBM = 98,99% (treino e teste)- 384 acertos e 96,11% (validação cruzada) : previsores - lgb.LGBMClassifier(num_leaves = 250, objective = 'binary', max_depth = 2, learning_rate = .05, max_bin =100)

CatBoost = 99,24% (treino e teste) previsores4 - 395 acertos e 97,16% (validação cruzada com previsores) - CatBoostClassifier(task_type='CPU', iterations=100, learning_rate=0.1, depth = 8, random_state = 5, eval_metric="Accuracy")

## **Transformando as classes strings em variáveis categóricas ordinais**

In [22]:
df2 = pd.DataFrame.copy(df)

In [23]:
df2['diagnosis'].replace({'M':1, 'B': 0}, inplace=True)

/tmp/ipykernel_23557/2475565386.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df2['diagnosis'].replace({'M':1, 'B': 0}, inplace=True)
/tmp/ipykernel_23557/2475565386.py:1: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df2['diagnosis'].replace({'M':1, 'B': 0}, inplace=True)


In [25]:
df2.head()

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0          1        17.99         10.38          122.80     1001.0   
1          1        20.57         17.77          132.90     1326.0   
2          1        19.69         21.25          130.00     1203.0   
3          1        11.42         20.38           77.58      386.1   
4          1        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280        198.0000              0.10430   

   symmetry_mean  ...  radius_worst  texture_worst  perimeter_worst  \
0         0.2419  ...         25.38          17.33           184.60   
1         0.1812  ...         24.99          23.41           158.80   
2         0.2069  ...         23.57          25.53           152.50   
3         0.2597  ...         14.91          26.50            98.87   
4         0.1809  ...         22.54          16.67           152.20   

   area_worst  smoothness_worst  compactness_worst  concavity_worst  \
0      2019.0            0.1622             0.6656           0.7119   
1      1956.0            0.1238             0.1866           0.2416   
2      1709.0            0.1444             0.4245           0.4504   
3       567.7            0.2098             0.8663           0.6869   
4      1575.0            0.1374           205.0000           0.4000   

   concave points_worst  symmetry_worst  fractal_dimension_worst  
0                0.2654          0.4601                  0.11890  
1              186.0000        275.0000                  0.08902  
2              243.0000          0.3613                  0.08758  
3                0.2575          0.6638                173.00000  
4                0.1625          0.2364                  0.07678  

[5 rows x 31 columns]

In [26]:
df2.dtypes

diagnosis                    int64
radius_mean                float64
texture_mean               float64
perimeter_mean             float64
area_mean                  float64
smoothness_mean            float64
compactness_mean           float64
concavity_mean             float64
concave points_mean        float64
symmetry_mean              float64
fractal_dimension_mean     float64
radius_se                  float64
texture_se                 float64
perimeter_se               float64
area_se                    float64
smoothness_se              float64
compactness_se             float64
concavity_se               float64
concave points_se          float64
symmetry_se                float64
fractal_dimension_se       float64
radius_worst               float64
texture_worst              float64
perimeter_worst            float64
area_worst                 float64
smoothness_worst           float64
compactness_worst          float64
concavity_worst            float64
concave points_worst

In [27]:
df2.shape

(569, 31)

## **ATRIBUTOS PREVISORES E ALVO**

In [75]:
previsores = df2.iloc[:, 1:31].values


In [76]:
previsores

array([[1.799e+01, 1.038e+01, 1.228e+02, ..., 2.654e-01, 4.601e-01,
        1.189e-01],
       [2.057e+01, 1.777e+01, 1.329e+02, ..., 1.860e+02, 2.750e+02,
        8.902e-02],
       [1.969e+01, 2.125e+01, 1.300e+02, ..., 2.430e+02, 3.613e-01,
        8.758e-02],
       ...,
       [1.660e+01, 2.808e+01, 1.083e+02, ..., 1.418e-01, 2.218e-01,
        7.820e-02],
       [2.060e+01, 2.933e+01, 1.401e+02, ..., 2.650e+02, 4.087e-01,
        1.240e+02],
       [7.760e+00, 2.454e+01, 4.792e+01, ..., 0.000e+00, 2.871e-01,
        7.039e-02]])

In [38]:
previsores.shape

(569, 30)

In [41]:
alvo = df2.iloc[:, 0].values

In [42]:
alvo

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1,
       1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 1,
       0, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1, 1,
       0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0,
       0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 0, 1,
       1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0,
       0, 1, 0, 0, 1, 1, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 1,
       1, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 0, 1, 0, 1, 1, 1, 1, 0, 0, 1, 1,
       0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 0,
       0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1,

In [43]:
alvo.shape

(569,)

## **Análise das escalas dos atributos (Escalonamento)**

In [44]:
df2.describe()

diagnosis  radius_mean  texture_mean  perimeter_mean    area_mean  \
count  569.000000   569.000000    569.000000      569.000000   569.000000   
mean     0.372583   706.771388     19.289649       91.969033   654.889104   
std      0.483918  2430.243368      4.301036       24.298981   351.914129   
min      0.000000     7.760000      9.710000       43.790000   143.500000   
25%      0.000000    12.210000     16.170000       75.170000   420.300000   
50%      0.000000    13.850000     18.840000       86.240000   551.100000   
75%      1.000000    17.680000     21.800000      104.100000   782.700000   
max      1.000000  9904.000000     39.280000      188.500000  2501.000000   

       smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
count       569.000000        569.000000      569.000000           569.000000   
mean          4.304801          4.835984        7.489124             2.366459   
std          21.074558         26.827478       35.618994            16.155145   
min           0.052630          0.019380        0.000000             0.000000   
25%           0.086410          0.065260        0.029580             0.020310   
50%           0.095940          0.094620        0.063870             0.033900   
75%           0.106100          0.132500        0.142500             0.077260   
max         123.000000        277.000000      313.000000           162.000000   

       symmetry_mean  ...  radius_worst  texture_worst  perimeter_worst  \
count     569.000000  ...    569.000000     569.000000       569.000000   
mean       16.965766  ...    315.194921      25.677223       107.261213   
std        53.846023  ...   1655.459336       6.146258        33.602542   
min         0.116700  ...      7.930000      12.020000        50.410000   
25%         0.163400  ...     13.180000      21.080000        84.110000   
50%         0.181400  ...     15.150000      25.410000        97.660000   
75%         0.203600  ...     19.850000      29.720000       125.400000   
max       304.000000  ...   9981.000000      49.540000       251.200000   

        area_worst  smoothness_worst  compactness_worst  concavity_worst  \
count   569.000000        569.000000         569.000000       569.000000   
mean    880.583128         10.633281          25.259112        26.723742   
std     569.356993         37.236433          96.473015       114.204035   
min     185.200000          0.071170           0.027290         0.000000   
25%     515.300000          0.117800           0.150700         0.116800   
50%     686.500000          0.133800           0.227900         0.249200   
75%    1084.000000          0.150000           0.384200         0.431600   
max    4254.000000        185.000000        1058.000000      1252.000000   

       concave points_worst  symmetry_worst  fractal_dimension_worst  
count            569.000000      569.000000               569.000000  
mean               8.745685       30.367174                 1.964313  
std               39.465975       90.748044                14.464355  
min                0.000000        0.156500                 0.055040  
25%                0.064990        0.254900                 0.071460  
50%                0.101500        0.288400                 0.080060  
75%                0.170800        0.331800                 0.092110  
max              291.000000      544.000000               173.000000  

[8 rows x 31 columns]

Padronização (utiliza a média e o desvio padrão como referência).

Normalização (utiliza os valores máximo e mínimo como referência).

In [45]:
from sklearn.preprocessing import StandardScaler

In [46]:
previsores_esc = StandardScaler().fit_transform(previsores)

In [47]:
previsores_esc

array([[-0.28367013, -2.07333501,  1.26993369, ..., -0.21506493,
        -0.32985162, -0.12769577],
       [-0.28260757, -0.35363241,  1.68595471, ...,  4.49527168,
         2.6981086 , -0.12976335],
       [-0.28296999,  0.45618695,  1.56650313, ...,  5.94082456,
        -0.33094131, -0.129863  ],
       ...,
       [-0.28424259,  2.0455738 ,  0.67267578, ..., -0.21819949,
        -0.33247988, -0.13051206],
       [-0.28259521,  2.33645719,  1.98252415, ...,  6.49875725,
        -0.33041852,  8.44441856],
       [-0.28788329,  1.22179204, -1.81438851, ..., -0.22179562,
        -0.33175967, -0.13105248]])

In [48]:
previsoresdf = pd.DataFrame(previsores_esc)
previsoresdf

0         1         2         3         4          5         6   \
0   -0.283670 -2.073335  1.269934  0.984375 -0.198822  -0.170064 -0.202009   
1   -0.282608 -0.353632  1.685955  1.908708 -0.200421  -0.177487 -0.208000   
2   -0.282970  0.456187  1.566503  1.558884 -0.199240  -0.174455 -0.204895   
3   -0.286376  0.253732 -0.592687 -0.764464 -0.197677  -0.169829 -0.203658   
4   -0.282723 -1.151816  1.776573  1.826229 -0.199682  -0.175466  5.353282   
..        ...       ...       ...       ...       ...        ...       ...   
564 -0.282200  0.721473  2.060786  2.343856  5.067204  -0.176097 -0.203588   
565 -0.282789  2.085134  1.615931  1.723842 -0.199800  -0.176563  3.835903   
566 -0.284243  2.045574  0.672676  0.577953 -0.200430  -0.176604 -0.207842   
567 -0.282595  2.336457  1.982524  1.735218 -0.198850  10.153899 -0.200567   
568 -0.287883  1.221792 -1.814389 -1.347789 -0.201946  -0.178794 -0.210441   

           7         8         9   ...        20        21        22  \
0   -0.137499 -0.310860 -0.108831  ... -0.175220 -1.359293  2.303601   
1   -0.142265 -0.311988 -0.111937  ... -0.175456 -0.369203  1.535126   
2   -0.138688 -0.311511 -0.111469  ... -0.176315 -0.023974  1.347475   
3   -0.140095 -0.310529 -0.106192  ... -0.181550  0.133984 -0.249939   
4   -0.140150 -0.311994 -0.111632  ... -0.176937 -1.466770  1.338539   
..        ...       ...       ...  ...       ...       ...       ...   
564 -0.138007 -0.312148 -0.111999  ... -0.175178  0.117700  1.752563   
565 -0.140546 -0.312100 -0.112125  ... -0.176242  2.047399  1.421940   
566 -0.143327  2.640106 -0.111963  ... -0.179090  1.374854  0.579001   
567  9.270434 -0.310901 -0.110036  ... -0.175003  2.237926  2.303601   
568 -0.146612 -0.312407 -0.111631  ...  5.526471  0.764190 -1.432735   

           23        24        25        26        27        28         29  
0    2.001237 -0.281453 -0.255151 -0.227967 -0.215065 -0.329852  -0.127696  
1    1.890489 -0.282485 -0.260120 -0.232089  4.495272  2.698109  -0.129763  
2    1.456285 -0.281931 -0.257652 -0.230259  5.940825 -0.330941  -0.129863  
3   -0.550021 -0.280173 -0.253068 -0.228186 -0.215265 -0.327605  11.835037  
4    1.220724 -0.282119  1.864760 -0.230700 -0.217675 -0.332319  -0.130610  
..        ...       ...       ...       ...       ...       ...        ...  
564  2.015301  3.504133 -0.259864 -0.230607 -0.216176  1.937093  -0.131000  
565  1.494959 -0.282678 -0.260062 -0.231388 -0.217667 -0.332089  -0.131331  
566  0.427906 -0.282751 -0.258846 -0.231224 -0.218199 -0.332480  -0.130512  
567  1.653171  4.149231 -0.253050 -0.225979  6.498757 -0.330419   8.444419  
568 -1.075813 -0.283394 -0.261388 -0.234206 -0.221796 -0.331760  -0.131052  

[569 rows x 30 columns]

In [49]:
previsoresdf.describe()

0             1             2             3             4   \
count  5.690000e+02  5.690000e+02  5.690000e+02  5.690000e+02  5.690000e+02   
mean   4.682839e-17  6.868164e-17 -1.248757e-16 -2.185325e-16  2.809703e-17   
std    1.000880e+00  1.000880e+00  1.000880e+00  1.000880e+00  1.000880e+00   
min   -2.878833e-01 -2.229249e+00 -1.984504e+00 -1.454443e+00 -2.019455e-01   
25%   -2.860506e-01 -7.259631e-01 -6.919555e-01 -6.671955e-01 -2.003412e-01   
50%   -2.853752e-01 -1.046362e-01 -2.359800e-01 -2.951869e-01 -1.998886e-01   
75%   -2.837978e-01  5.841756e-01  4.996769e-01  3.635073e-01 -1.994061e-01   
max    3.787819e+00  4.651889e+00  3.976130e+00  5.250529e+00  5.637112e+00   

                 5             6             7             8             9   \
count  5.690000e+02  5.690000e+02  5.690000e+02  5.690000e+02  5.690000e+02   
mean   9.365678e-18 -5.931596e-17  3.121893e-18 -1.873136e-17  3.121893e-18   
std    1.000880e+00  1.000880e+00  1.000880e+00  1.000880e+00  1.000880e+00   
min   -1.796979e-01 -2.104415e-01 -1.466122e-01 -3.131873e-01 -1.128821e-01   
25%   -1.779862e-01 -2.096103e-01 -1.453539e-01 -3.123193e-01 -1.117774e-01   
50%   -1.768909e-01 -2.086468e-01 -1.445119e-01 -3.119847e-01 -1.112336e-01   
75%   -1.754777e-01 -2.064373e-01 -1.418256e-01 -3.115720e-01 -1.105657e-01   
max    1.015390e+01  8.584737e+00  9.889976e+00  5.335339e+00  1.087025e+01   

       ...            20            21            22          23  \
count  ...  5.690000e+02  5.690000e+02  5.690000e+02  569.000000   
mean   ...  4.682839e-18  1.248757e-17 -3.746271e-16    0.000000   
std    ...  1.000880e+00  1.000880e+00  1.000880e+00    1.000880   
min    ... -1.857704e-01 -2.223994e+00 -1.693361e+00   -1.222423   
25%    ... -1.825962e-01 -7.486293e-01 -6.895783e-01   -0.642136   
50%    ... -1.814052e-01 -4.351564e-02 -2.859802e-01   -0.341181   
75%    ... -1.785636e-01  6.583411e-01  5.402790e-01    0.357589   
max    ...  5.843883e+00  3.885905e+00  4.287337e+00    5.930172   

                 24            25            26            27            28  \
count  5.690000e+02  5.690000e+02  5.690000e+02  5.690000e+02  5.690000e+02   
mean   1.248757e-17  1.248757e-17 -6.243785e-18  1.873136e-17  2.965798e-17   
std    1.000880e+00  1.000880e+00  1.000880e+00  1.000880e+00  1.000880e+00   
min   -2.838995e-01 -2.617729e-01 -2.342059e-01 -2.217956e-01 -3.332001e-01   
25%   -2.826461e-01 -2.604926e-01 -2.331822e-01 -2.201474e-01 -3.321148e-01   
50%   -2.822161e-01 -2.596917e-01 -2.320219e-01 -2.192215e-01 -3.317453e-01   
75%   -2.817806e-01 -2.580701e-01 -2.304234e-01 -2.174640e-01 -3.312667e-01   
max    4.686812e+00  1.071439e+01  1.073828e+01  7.158132e+00  5.664968e+00   

                 29  
count  5.690000e+02  
mean  -3.746271e-17  
std    1.000880e+00  
min   -1.321146e-01  
25%   -1.309784e-01  
50%   -1.303834e-01  
75%   -1.295495e-01  
max    1.183504e+01  

[8 rows x 30 columns]

## **RESUMO PRÉ-PROCESSAMENTO**

alvo = variável que se pretende atingir (tem ou não doença cardíaca).

previsores = conjunto de variáveis previsoras com as variáveis categóricas transformadas em numéricas manualmente, sem escalonar.

previsores_esc = conjunto de variáveis previsoras com as variáveis categóricas transformadas em numéricas, escalonada

## **BASE DE TREINO E TESTE**

In [63]:
from sklearn.model_selection import train_test_split

In [151]:
x_treino, x_teste, y_treino, y_teste = train_test_split(previsores_esc, alvo, test_size = 0.3, random_state = 0)

In [128]:
x_treino.shape

(398, 30)

In [129]:
x_teste.shape

(171, 30)

In [130]:
y_treino.shape

(398,)

In [131]:
y_teste.shape

(171,)

# **CRIAÇÃO DO ALGORITMO**

In [132]:
from sklearn.neural_network import MLPClassifier

In [133]:
MLPClassifier?

Init signature:
MLPClassifier(
    hidden_layer_sizes=(100,),
    activation='relu',
    *,
    solver='adam',
    alpha=0.0001,
    batch_size='auto',
    learning_rate='constant',
    learning_rate_init=0.001,
    power_t=0.5,
    max_iter=200,
    shuffle=True,
    random_state=None,
    tol=0.0001,
    verbose=False,
    warm_start=False,
    momentum=0.9,
    nesterovs_momentum=True,
    early_stopping=False,
    validation_fraction=0.1,
    beta_1=0.9,
    beta_2=0.999,
    epsilon=1e-08,
    n_iter_no_change=10,
    max_fun=15000,
)
Docstring:     
Multi-layer Perceptron classifier.

This model optimizes the log-loss function using LBFGS or stochastic
gradient descent.

.. versionadded:: 0.18

Parameters
----------
hidden_layer_sizes : array-like of shape(n_layers - 2,), default=(100,)
    The ith element represents the number of neurons in the ith
    hidden layer.

activation : {'identity', 'logistic', 'tanh', 'relu'}, default='relu'
    Activation function for the hidden laye

### Parâmetros MLPClassifier   
- hidden_layer_sizes (camadas escondidas): default (100,)   

  Quant.= (Ne+Ns)/2 = (31+1)/2 = 16 neurônios   
  Quant.=2/3.(Ne) + Ns = 2/3.31+1 = 21 neurônios
  
- activation: Função de ativação default='relu' 
- solver: algoritmo matemático. Default='adam' (datasets grandes = acima de 1000 amostras). lbfgs é para datasets pequenos. sgd é com a descida do gradiente estocástico (recomendado testar).
- alpha:  parâmetro para o termo de regularização de ajuste de pesos. Aumento de alpha estimula pesos menores e diminuição de alpha estimula pesos maiores. Default=0.0001.
- batch_size: tamanho dos mini lotes. default=min(200, n_samples). Não usar com o solver lbfgs.
- learning_rate: taxa de aprendizagem. default='constant'. Três tipos:   
  1- 'constant':uma taxa de aprendizado constante dada pela taxa de aprendizagem inicial.   
  2- 'invscaling': diminui gradualmente por: taxa efetiva = taxa inicial / t^power_t  
  3- 'adaptive': a taxa é dividida por 5 cada vez que em duas épocas consecutivas não diminuir o erro.
- learning_rate_init: taxa de aprendizagem inicial. Default=0.001
- max_iter int: Número máximo de iterações. default = 200.  ('sgd', 'adam').
- max_fun: Número máximo de chamadas de função de perda. Para 'lbfgs'. Default: 15000
- shuffle: default = True
Usado apenas quando solver = 'sgd' ou 'adam'.
- random_state: default = None
- tol:Tolerância para a otimização.Default=0.0001
- momentum: otimização do algoritmo 'sgd'. Default: 0.9.
- n_iter_no_change: Número máximo de épocas que não atinge a tolerância de melhoria. default = 10. Apenas para solver = 'sgd' ou 'adam'
- verbose : Mostra o progresso. default=False.



In [152]:
redes = MLPClassifier(hidden_layer_sizes=(18), activation='relu', solver='adam', max_iter =800,
                              tol=0.0001, random_state = 3, verbose = True)
                               

In [153]:
redes.fit(x_treino, y_treino)

Iteration 1, loss = 1.04418961
Iteration 2, loss = 1.01670327
Iteration 3, loss = 0.98981291
Iteration 4, loss = 0.96383641
Iteration 5, loss = 0.93909256
Iteration 6, loss = 0.91461914
Iteration 7, loss = 0.89099079
Iteration 8, loss = 0.86719805
Iteration 9, loss = 0.84546158
Iteration 10, loss = 0.82402246
Iteration 11, loss = 0.80273179
Iteration 12, loss = 0.78227581
Iteration 13, loss = 0.76270026
Iteration 14, loss = 0.74448876
Iteration 15, loss = 0.72609533
Iteration 16, loss = 0.70867251
Iteration 17, loss = 0.69181875
Iteration 18, loss = 0.67579957
Iteration 19, loss = 0.66008707
Iteration 20, loss = 0.64503395
Iteration 21, loss = 0.63032947
Iteration 22, loss = 0.61599651
Iteration 23, loss = 0.60261044
Iteration 24, loss = 0.58955940
Iteration 25, loss = 0.57692302
Iteration 26, loss = 0.56477284
Iteration 27, loss = 0.55274751
Iteration 28, loss = 0.54132619
Iteration 29, loss = 0.53043755
Iteration 30, loss = 0.51973902
Iteration 31, loss = 0.50951844
Iteration 32, los

MLPClassifier(hidden_layer_sizes=18, max_iter=800, random_state=3, verbose=True)

In [154]:
previsoes = redes.predict(x_teste)

In [155]:
previsoes

array([1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 1, 1, 1,
       0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0,
       1, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0,
       1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 1, 0, 0, 0,
       1, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1,
       0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0])

In [156]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

In [157]:
print("Acurácia: %.2f%%" % (accuracy_score(y_teste, previsoes) * 100.0))

Acurácia: 93.57%


In [158]:
confusion_matrix(y_teste, previsoes)

array([[101,   7],
       [  4,  59]])

In [159]:
print(classification_report(y_teste, previsoes))

              precision    recall  f1-score   support

           0       0.96      0.94      0.95       108
           1       0.89      0.94      0.91        63

    accuracy                           0.94       171
   macro avg       0.93      0.94      0.93       171
weighted avg       0.94      0.94      0.94       171



**Análise dados de treino**

In [160]:
previsoes_treino = redes.predict(x_treino)
previsoes_treino

array([0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 1,
       0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1,
       0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0,
       1, 0, 0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1,
       0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0,
       0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0,
       1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0,
       0, 0, 1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 1,
       0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1,
       0, 0, 1, 1, 1, 1, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1,
       0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 1,
       1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0,

In [161]:
accuracy_score(y_treino, previsoes_treino)

0.9824120603015075

In [162]:
confusion_matrix(y_treino, previsoes_treino)

array([[248,   1],
       [  6, 143]])

### **Validação Cruzada**

In [163]:
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score

In [164]:
# Separando os dados em folds
kfold = KFold(n_splits = 30, shuffle=True, random_state = 5)

In [165]:
# Criando o modelo
modelo = MLPClassifier(hidden_layer_sizes=(18), activation='relu', solver='adam', max_iter =800,
                              tol=0.0001, random_state = 3, verbose = True)
resultado = cross_val_score(modelo, previsores, alvo, cv = kfold)
print("Acurácia Média: %.2f%%" % (resultado.mean() * 100.0))

Iteration 1, loss = 21.03635875
Iteration 2, loss = 20.69925693
Iteration 3, loss = 20.65234790
Iteration 4, loss = 20.66590316
Iteration 5, loss = 20.48508250
Iteration 6, loss = 19.78094703
Iteration 7, loss = 18.07803764
Iteration 8, loss = 16.34181129
Iteration 9, loss = 13.79892320
Iteration 10, loss = 11.90110900
Iteration 11, loss = 11.03005936
Iteration 12, loss = 10.24037785
Iteration 13, loss = 9.64378096
Iteration 14, loss = 9.33705057
Iteration 15, loss = 8.99351581
Iteration 16, loss = 8.76159624
Iteration 17, loss = 8.63414272
Iteration 18, loss = 8.52010148
Iteration 19, loss = 8.47063257
Iteration 20, loss = 8.34084576
Iteration 21, loss = 8.26153410
Iteration 22, loss = 8.15139699
Iteration 23, loss = 7.96463853
Iteration 24, loss = 7.75390264
Iteration 25, loss = 7.50159806
Iteration 26, loss = 7.18803969
Iteration 27, loss = 6.90758410
Iteration 28, loss = 6.61466799
Iteration 29, loss = 6.35030593
Iteration 30, loss = 6.09753709
Iteration 31, loss = 5.89137011
Itera

Naive Bayes = 84,78% (treino e teste) - 234 acertos e 85,17% (validação cruzada) : previsores3_esc

SVM = 86,23% - 238 acertos e 85,72% (validação cruzada) : previsores3-esc - SVC(kernel='rbf', random_state=1, C = 2)

Regressão logística = 86,23% - 238 acertos e 85,83% (validação cruzada) : previsores3_esc - LogisticRegression(random_state=1, max_iter=600, penalty="l2", tol=0.0001, C=1,solver="lbfgs")

KNN = 84,06% - 232 acertos e 85,84% (validação cruzada) : previsores3_esc ou previsores_esc - KNeighborsClassifier(n_neighbors=7, metric='minkowski', p = 1)

Árvore de decisão = 83,7% (treino e teste) - 231 acertos e 83,32% (validação cruzada): previsores3_esc - DecisionTreeClassifier(criterion='entropy', random_state = 0, max_depth=3)

Random Forest = 86,59% - 239 acertos e 85,92% (validação cruzada) : previsores - RandomForestClassifier(n_estimators=150, criterion='entropy', random_state = 0, max_depth=4)

**XGboost = 87,68% - 242 acertos e 87,13% (validação cruzada) : previsores - XGBClassifier(max_depth=2, learning_rate=0.05, n_estimators=250, objective='binary:logistic', random_state=3)**

LightGBM = 86,23% (treino e teste)- 238 acertos e 85,93% (validação cruzada) : previsores - lgb.LGBMClassifier(num_leaves = 250, objective = 'binary',  max_depth = 2, learning_rate = .05, max_bin =100)

CatBoost = 86,96% (treino e teste) previsores4 - 240 acertos e 88% (validação cruzada com previsores) - CatBoostClassifier(task_type='CPU', iterations=100, learning_rate=0.1, depth = 8, random_state = 5, eval_metric="Accuracy")

Redes Neurais Artificiais = 93.57% (treino e teste) - 160 acertos e 89.78% (validação cruzada) : previsores-esc - MLPClassifier(hidden_layer_sizes=(18), activation='relu', solver='adam', max_iter =800,
                              tol=0.0001, random_state = 3, verbose = True)